inception-recommendation-api/src/main/resources/META-INF/asciidoc/developer-guide/implementing_recommender_tutorial.adoc

```diff
@@ -0,0 +1,362 @@
+predicting, i.e. it predicts always the label that appears most often in the training data. The full
+code for this example can be found on https://github.com/inception-project/inception/inception-imls-data-majority[on Github].
+
+== Setting up the environment
+
+We assume that you have your own fork of INCEpTION and want to add a new internal recommender to it.
+To get started, check out the most recent source code of INCEpTION from
+https://github.com/inception-project/inception[Github] and import it as a Maven project in the IDE
+of your choice. Add a new project to the workspace of your IDE (do not add a new module to the
+INCEpTION project itself), we will call it `inception-imls-data-majority`.
+
+In the pom of the INCEpTION project, add your recommender as a dependency.
+
+....
+<dependencies>
+…
+    <dependency>
+        <groupId>de.tudarmstadt.ukp.inception.app</groupId>
+        <artifactId>inception-imls-data-majority</artifactId>
+        <version>0.1.0-SNAPSHOT</version>
+    </dependency>
+…
+</dependencies>
+....
+
+Do the same in `inception-app-webapp`, but omit the version number. It then uses automatically the
+version in the parent pom. Also add it to `usedDependencies` there.
+
+To add a new recommender to INCEpTION, at least two interfaces need to be implemented. These are
+described in the following.
+
+== Implementing the RecommendationEngine
+
+Recommenders give suggestions for possible annotations to the user. In order to do that,
+they need to be able be to trained on existing annotations, predict annotations in a document and
+be evaluated for a performance estimate. This is what the https://github.com/inception-project/inception/blob/master/inception-recommendation-api/src/main/java/de/tudarmstadt/ukp/inception/recommendation/api/recommender/RecommendationEngine.java[RecommendationEngine] interface is for.
+It defines the methods that are used to train, test and evaluate a machine learning algorithm.
+Instances of this interface often wrap external machine learning packages like _OpenNLP_ or
+_Deeplearning4j_.
+
+Recommenders in INCEpTION heavily rely on https://uima.apache.org[Apache UIMA] types and features.
+A recommender is configured for a certain layer and a certain feature. A layer can be seen as the
+type of annotation you want to to, e.g. `POS`, `NER`. Layers correspond to UIMA types. A feature is
+one piece of information that should be annotated, e.g. the POS tag. One layer can have many features.
+
+Annotations are given to a recommender in the form of a
+http://uima.apache.org/d/uimaj-current/apidocs/index.html[UIMA CAS]. One CAS corresponds to one
+document in INCEpTION. Annotations from a CAS can be read and manipulated via the
+https://uima.apache.org/d/uimafit-current/api/org/apache/uima/fit/util/CasUtil.html[CasUtil].
+
+We start by creating a new class `DataMajorityNerRecommender` that implements `RecommendationEngine`.
+Please see the JavaDoc of the respective methods for their semantics.
```

Comment:
I see no JavaDoc in the example below. No comments at all in fact ;) Some comments would be helpful.